posix: syslog: add support for syslog

cfriedt opened the pull request:
This is just a simple wrapper around Zephyr's native log api.

See https://pubs.opengroup.org/onlinepubs/9699919799/functions/closelog.html for additional information.

Commit:
posix: syslog: add support for syslog

This is just a simple wrapper around Zephyr's native log api.

Note: the standard LOG_ERR syslog priority does conflict with
Zephyr's LOG_ERR() macro. This will need to be worked-around on
a case-by-case basis.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented on lib/posix/options/syslog.c:
```diff
@@ -0,0 +1,89 @@
+#include <zephyr/logging/log.h>
+#include <zephyr/posix/syslog.h>
+
+#define _LOG_ERR 3
```
Likely not

ycsin commented on lib/posix/options/syslog.c:
```diff
@@ -0,0 +1,87 @@
+static struct k_spinlock syslog_lock;
+static uint8_t syslog_mask;
+
+static inline int syslog_priority_to_zephyr_log_level(uint8_t priority)
```
`priority` was `int` in `vsyslog`

ycsin commented on samples/posix/syslog/src/main.c:
```diff
@@ -0,0 +1,29 @@
+	openlog("main", LOG_PID | LOG_CONS | LOG_NOWAIT, LOG_LOCAL7);
+	(void) setlogmask(LOG_MASK(-1));
+
+	ARRAY_FOR_EACH(prios, i) {
```
Not sure if it makes sense to use typical for-loop so that this sample is truly portable and can be built on host OS as well

cfriedt replied:
Yeah, makes sense.

cfriedt commented on lib/posix/options/syslog.c:
```diff
@@ -0,0 +1,87 @@
+		return;
+	}
+
+	log2_generic(level, format, ap);
```
@nordic-krch - is there a particular Kconfig option I should check for here? If the `LOG_CONS` option is specified and logging is expected to be a no-op, I should probably use `printk()` instead.

cfriedt replied:
I think this could do the trick. Let me know if it works for you.

nordic-krch replied:
don't you want to add level when minimal logging is used?

I thought i removed all remains of v2 naming in logging but somehow that was left.

cfriedt replied:
@nordic-krch - I hope it's ok, I prepended a commit to deprecate `log2_generic()` and added `log_generic()` (basically just a rename).

It's possible that log level, time, and other items could be added in the future, but I would prefer to avoid multiple calls to `(v)printk()` at this point because it's possible that messages might not be printed atomically.

nordic-krch replied:
>  I prepended a commit to deprecate log2_generic()

Thanks for doing that!

ycsin commented on samples/posix/syslog/src/main.c:
```diff
@@ -0,0 +1,33 @@
+	closelog();
+
+	/* yield briefly to logging thread */
+	sleep(100);
```
1 second should be enough for the logging thread to finish its work

```suggestion
	sleep(1);
```

ycsin reviewed (dismissed):
LGTM, would be nice to add a makefile to build directly for host OS

cfriedt commented on samples/posix/syslog/src/main.c:
```diff
@@ -0,0 +1,33 @@
+	closelog();
+
+	/* yield briefly to logging thread */
+	sleep(100);
```
Lol, good catch - I originally had 1s and then wanted to go to msleep. Jetlag 😅

nordic-krch reviewed (dismissed):
looks good, i minor comment that may be applied.

kartben commented on samples/posix/syslog/README.rst:
```diff
@@ -0,0 +1,38 @@
+Overview
+********
+
+In this sample application, the POSIX `syslog()`_ function is used to log a messages to each of
+the startdard priority levels.
```
```suggestion
In this sample application, the POSIX `syslog()`_ function is used to log a message to each of
the standard priority levels.
```

kartben commented on samples/posix/syslog/src/main.c:
```diff
@@ -0,0 +1,33 @@
+
+#define N_PRIOS 8
+
+int main(void)
+{
+	int prios[N_PRIOS] = {
+		LOG_EMERG,   LOG_ALERT,  LOG_CRIT, LOG_ERR,
+		LOG_WARNING, LOG_NOTICE, LOG_INFO, LOG_DEBUG,
+	};
+
+	openlog("main", LOG_PID | LOG_CONS | LOG_NOWAIT, LOG_LOCAL7);
+	(void)setlogmask(LOG_MASK(-1));
+
+	for (size_t i = 0; i < N_PRIOS; ++i) {
+		syslog(i, "syslog priority %d", prios[i]);
+	}
+
+	closelog();
+
+	/* yield briefly to logging thread */
+	usleep(100000);
+
+	return 0;
+}
```
Not necessarily blocking, but important nevertheless :)

I am starting to think it probably doesn't scale all that well to keep adding samples that feel like they are "just" exercising POSIX APIs, and not really showing anything new to the end user. "Of course" syslog should just work, if Zephyr is POSIX compliant, so is there really a point in showing it, especially if no Zephyr-specific tweak is needed?

Surely, having *one* minimal POSIX sample is necessary so that people can quickly get a sense of the Kconfigs they need, or anything else needed to get their code "POSIX-ready", is needed, but not sure we need more of such "minimal" samples (plus, they are costly to run in CI -- and probably paraphrasing existing tests anyway?)

It would probably make more sense to have a much richer sample that effectively uses tons of POSIX APIs to make the point that It Just Works™, once and for all.

cfriedt replied:
@kartben - I could potentially move the sample to the existing tests/posix/common suite. As long as it compiles and runs that should be sufficient. Does that work for you?

kartben replied:
Yup, that would definitely work!

Commit:
tests: posix: add a compile-and-run-only test for syslog

Ensure that we have code coverage for syslog and that it
can compiles and run with regular tests.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented:
@kartben - please let me know if this latest rev addresses your change request.

Also, I wonder if this PR should be considered for inclusion in v3.6.0 because it fixes an api that should have been deprecated / renamed some time ago in logging (`log2_generic()` -> `log_generic()`).

cfriedt commented:
Moved the `log_generic()` change to #68949

cfriedt commented:
Rebased on #68949 (`DNM` until after release)

cfriedt commented:
@nordic-krch - does this change look OK to you?
